### DISCUSSION TITLE
Default value + object check

### DISCUSSION THREAD

**Discussion by hagen31** [sadieblue/agilityblue.workspace-scripts#18] [Q&A] [answered]
Is it possible to have a default value appear on a field when a form is opened? For example, I want to have the Matter Reference value populate in the MediaLogEntry Reference field when a user creates a new Media Log Entry. I have it setup to populate the value; however, it does not do it until the entry is saved. I would like the user to be able to see the value is already populated.

Additionally, is it possible to loop through the MediaLogEntry Reference values for the matter and verify the new value is unique. In an ideal situation I would like the system to automatically populate the next value incrementing by 1.

**Comment by mrabb24** [comment]
Hi,

Q1: There's no hook from the UI to tell a script that some UI action is being performed and scripts can only alter the behavior of items before or after saving them, so it's not possible to have a value populate a form as the result of opening a form window. 

Q2: Yes, but there would likely be a more efficient way to do this vs looping through each value hopefully, otherwise the script may take too long to run and time out after a minute. Does the number you wish to auto-increment just need to look at the last highest number of that field and increment by 1? I may have an example of that, if that's the case.

Matt

**Reply by hagen31** [reply]
Thank for Q1 that matches what I was finding was just hoping there was a way.

Q2 - Yes that is essentially what I want to do; however, the field is a combination of text and number and must only increment for that matter. For example, I could have a media log reference of ABCD_0001 and if a user went to create a new media on VXYZ I would want that to be VXYZ_0001. Then if a user was on ABCD it would be ABCD_0002 as 0002 is the next available for that matter.

**Reply by mrabb24** [reply]
Thanks. I don't have any examples that are this detailed, but it is certainly possible. Where is the prefix for a media log entry defined for the matter? How would the script come up with the "ABCD" portion of the number?

**Reply by hagen31** [reply]
It pulls the Matter Reference value on the Media Log Entry object. The number portion is four digits incrementing by 1.

**Comment by mrabb24** [comment]
You can get the matter reference by retrieving the matter based on the matter id that a media log entry is being saved to. To find the last number for a specific matter, you could filter on the media log entries collection for entries that start with the matter reference, then you would order by the auto incrementing field in descending order and retrieve only 1 record. This would get you the last media log entry that used that matter's auto incrementing number and is more efficient over trying to iterate over all the media log entries.

I modified my example to work out unique increments on the media log entry reference field based on the matter. This should get you either everything you need, or very close at least:

```powershell
$mediaLogEntry = Get-InboundObjectInstance

if (-not ($null -eq $mediaLogEntry.MatterId)) 
{

    # The media log entry is being associated to a matter
    
    $matter = Get-Matter -Id $mediaLogEntry.MatterId
    
    
    # This is a safety check just to make sure that the matter has a reference
    if (-not ($null -eq $matter.Reference -and $matter.Reference.Length -eq 0)) 
    {
        
        Write-Output "Has matter reference: $($matter.Reference)"
        
        $lastMediaWithThisMatterRef = Get-MediaLogEntries `
            -Filter "startswith(Reference, '$($matter.Reference)')" `
            -OrderBy "Reference desc" `
            -Top 1
        
        if ($lastMediaWithThisMatterRef.TotalCount -gt 0)
        {
        
            # This means there was a media log entry that started with that matter ref.
            # Get the entry, and increment the ref by 1
            
            $previousRef = $lastMediaWithThisMatterRef.Collection[0].Reference
            
            # Split the string into prefix and numeric parts, assuming it is all separated
            # by an underscore
            $parts = $previousRef -split "_"
            $prefix = $parts[0]
            $number = [int]$parts[1]
            
            # Increment the numeric part
            $number++
            
            # Format the incremented number with leading zeros
            $newNumber = "{0:D4}" -f $number
            
            # Combine the prefix and the incremented number and save to the incoming ref
            $mediaLogEntry.Reference = "$($prefix)_$($newNumber)"
        
        }
        else
        {
        
            # This means that there are no entries that start with that matter ref, 
            # so this one should start at 1
            
            $mediaLogEntry.Reference = "$($matter.Reference)_0001"
        
        }
        
        # Output the result
        Write-Output "Saved reference as $($mediaLogEntry.Reference)"
        
    }
    
}

return $mediaLogEntry
```

Keep in mind that this is designed for media log entries that are directly associated with a matter. If you have media log entries associated to tasks or projects, you will need to handle those accordingly because those will not populate the `MatterId` field directly when saving the media log entry (they will have either a `ProjectId` or `TaskId`)

Matt

**Reply by hagen31** [reply]
Thanks I'll take a look at this. We usually associate our media log to a Project; however, have been considering switching that so I'll see if I can get this to work.

**Reply by hagen31** [reply]
This has worked well when associated to a Matter. Associating with a Task is proving more difficult as not all media is associated to the same task. With our workflow a new Project is opened up each time data is received along with a data tracking Task and a Processing Task. The media log is then associated with that task. The next time data comes in it would have a new Project/Task. Any ideas?

**Reply by mrabb24** [reply]
If you associate a media log entry with a task, then you will have a `TaskId` value on your media log entry. If you associate a media log entry with a project, then you will have a `ProjectId` value on your media log entry. You can chain your conditions to check for the existence of these values and retrieve the appropriate object to get to the matter reference.  This means you will need to approach the script differently. Here is an example of checking each type first, starting at the lowest level:

```powershell
if (-not ($null -eq $mediaLogEntry.TaskId)) 
{
    $matterReference = (Get-Task -Id $mediaLogEntry.TaskId).MatterReference
}

if (-not ($null -eq $mediaLogEntry.ProjectId)) 
{
    $matterReference = (Get-Project -Id $mediaLogEntry.ProjectId).MatterReference
}

if (-not ($null -eq $mediaLogEntry.MatterId)) 
{
    $matterReference = (Get-Matter -Id $mediaLogEntry.MatterId).Reference
}

# Continue on with your script...
if (-not ($null -eq $matterReference -and $matterReference.Length -eq 0)) 
{
    # etc...
}
```

Matt

**Reply by hagen31** [reply]
I was getting a null return so to back it up I went all the way to just 

if (-not ($null -eq $mediaLogEntry.ProjectId)) 
{
    $matterReference = (Get-Project -Id $mediaLogEntry.ProjectId).MatterReference
}

Write-Output "Has matter reference: $($matterReference)"

I then received this error.
Has matter reference: 
[Uncaught Exception] Cannot bind argument to parameter 'Id' because it is null.

**Reply by bmward80** [reply]
Hi Austin, Matt is out today, but I can probably help with this. You can try breaking up the shorthand notation into a couple of lines, like this:

```powershell
if (-not ($null -eq $mediaLogEntry.ProjectId)) {
    $project = Get-Project -Id $mediaLogEntry.ProjectId
    $matterReference = $project.MatterReference
}
```

That might be better. Also, you can always add some debug output statements so you can see the data in your objects. For example, if something is complaining about being null, put a `ConvertTo-Json` call right before it so you can see what your object looks like prior to passing some data in:

```powershell
$mediaLogEntry | ConvertTo-Json
```

The output will show you every property on the object and you can look for the `ProjectId` property and see if it has any value. You may realize that something you're referencing is not being set somewhere.

Hope this helps,
Brandon

**Reply by hagen31** [reply, marked as answer]
Hello Brandon,

This was helpful. I've made some very slight changes and the below works. I am going to use this as a base to make some other field requirements.

```powershell
$mediaLogEntry = Get-InboundObjectInstance

if (-not ($null -eq $mediaLogEntry.TaskId)) {
    $task = Get-Task -Id $mediaLogEntry.TaskId
    $matterReference = $task.MatterReference
}

if (-not ($null -eq $mediaLogEntry.ProjectId)) {
    $project = Get-Project -Id $mediaLogEntry.ProjectId
    $matterReference = $project.MatterReference
}

if (-not ($null -eq $mediaLogEntry.MatterId)) {
    $matter = Get-Matter -Id $mediaLogEntry.MatterId
    $matterReference = $matter.Reference
}

if ($null -ne $matterReference -and $matterReference.Length -gt 0) {
    
    $lastMediaWithThisMatterRef = Get-MediaLogEntries `
        -Filter "startswith(Reference, '$($matterReference)')" `
        -OrderBy "Reference desc" `
        -Top 1
    
    
    if ($lastMediaWithThisMatterRef.TotalCount -gt 0) {
        $previousRef = $lastMediaWithThisMatterRef.Collection[0].Reference
        
        
        # Split the string into prefix and numeric parts, assuming it is all separated
        # by an underscore
        $parts = $previousRef -split "_"
        $prefix = $parts[0] + "_" + $parts[1]
        $number = [int]$parts[2]

        # Increment the numeric part
        $number++

        # Format the incremented number with leading zeros
        $newNumber = "{0:D4}" -f $number
        
        # Combine the prefix and the incremented number and save to the incoming ref
        $mediaLogEntry.Reference = "$($prefix)_$($newNumber)"
        
    } else {
        # This means that there are no entries that start with that matter ref,
        # so this one should start at 1
        $mediaLogEntry.Reference = "$($matterReference)_0001"
    }
    
    Write-Output "Saved reference as $($mediaLogEntry.Reference) for $($matterReference)."
} else {
    Write-Output "Matter reference is either null or empty."
    $null
}

return $mediaLogEntry
```
